openprescribing/frontend/models.py

```diff
@@ -448,6 +458,112 @@ def current_version(self):
+            if objs:
+                return objs
+        return []
+
+    @property
+    def dmd_obj_ids(self):
+        return [obj.id for obj in self.dmd_objs]
+
+    @property
+    def dmd_obj_type(self):
+        if not self.dmd_objs:
+            return None
+        else:
+            return type(self.dmd_objs[0])
+
+    def tariff_categories(self):
+        '''Return all tariff categories for this presentation.
+
+        Tariff categories are set at VMPP level, so this doesn't make sense for
+        presentations that correspond to AMPs.
+        '''
+
+        if self.dmd_obj_type == VMP:
+            filter_key = 'dtinfo__vmpp__vmp__id__in'
+        elif self.dmd_obj_type == VMPP:
+            filter_key = 'dtinfo__vmpp__id__in'
+        elif self.dmd_obj_type == AMPP:
+            filter_key = 'dtinfo__vmpp__ampp__id__in'
+        else:
+            return []
+
+        filter_params = {filter_key: self.dmd_obj_ids}
+        return DtPaymentCategory.objects.filter(**filter_params).distinct()
+
+    def tariff_categories_descr(self):
+        '''Return a description of the presentation's tariff category.'''
+        return ', '.join(tc.descr for tc in self.tariff_categories())
+
+    def availability_restrictions(self):
+        '''Return all availability restrictions for this presentation.
+
+        Availabiliy restrictions are set at AMP level, so this doesn't make
+        sense for presentations that correspond to VMPPs.
+        '''
+        if self.dmd_obj_type == VMP:
+            filter_key = 'amp__vmp__id__in'
+        elif self.dmd_obj_type == AMP:
+            filter_key = 'amp__id__in'
+        elif self.dmd_obj_type == AMPP:
+            filter_key = 'amp__ampp__id__in'
+        else:
+            return []
+
+        filter_params = {filter_key: self.dmd_obj_ids}
+        return AvailabilityRestriction.objects.filter(**filter_params).distinct()
+
+    def availability_restrictions_descr(self):
+        '''Return a description of the presentation's availabilty restriction.
+
+        If any AMPs have "None" as their availability restriction, we
+        consider that the presentation itself has no availability restriction.
+        '''
+        descrs = [ar.descr for ar in self.availability_restrictions()]
+        if 'None' in descrs:
+            return 'None'
+        else:
+            return ', '.join(descrs)
+
+    def prescribability_statuses(self):
+        '''Return all prescribability statuses for this presentation.
+        '''
+        if self.dmd_obj_type == VMP:
+            vmp_ids = self.dmd_obj_ids
+        elif self.dmd_obj_type in [AMP, VMPP]:
+            vmp_ids = set(obj.vmp_id for obj in self.dmd_objs)
+        elif self.dmd_obj_type == AMPP:
+            vmp_ids = set(ampp.amp.vmp_id for ampp in self.dmd_objs)
+        else:
+            return []
+
+        return VirtualProductPresStatus.objects.filter(vmp__id__in=vmp_ids).distinct()
```

Comment:
Does this lookup pattern (used a few places in this PR) end up doing an unnecessary join? Would this work equivalently:
```
return VirtualProductPresStatus.objects.filter(vmp__in=vmp_ids).distinct()
```
I don't suppose it matters much here in any case as it's not like the performance difference would be noticeable.

Reply:
They produce the same SQL:

```
>>> print(VirtualProductPresStatus.objects.filter(vmp__id__in=vmp_ids).distinct().query)
SELECT DISTINCT "dmd2_virtualproductpresstatus"."cd", "dmd2_virtualproductpresstatus"."descr" FROM "dmd2_virtualproductpresstatus" INNER JOIN "dmd2_vmp" ON ("dmd2_virtualproductpresstatus"."cd" = "dmd2_vmp"."pres_statcd") WHERE "dmd2_vmp"."vpid" IN (1, 2, 3)
>>> print(VirtualProductPresStatus.objects.filter(vmp__in=vmp_ids).distinct().query)
SELECT DISTINCT "dmd2_virtualproductpresstatus"."cd", "dmd2_virtualproductpresstatus"."descr" FROM "dmd2_virtualproductpresstatus" INNER JOIN "dmd2_vmp" ON ("dmd2_virtualproductpresstatus"."cd" = "dmd2_vmp"."pres_statcd") WHERE "dmd2_vmp"."vpid" IN (1, 2, 3)
```

Have I missed something?

Reply:
Ah, no. I have. Sorry, I misunderstood the relationship structure here.